panic: runtime error: index out of range [2091008] with length 2091008

Hit running the rac on some files I get a panic error. The rac file is `Payload MM_20210421-104323_20210422-13335.rac`

file can be found at: https://chalmersuniversity.box.com/s/s038kk7t0xyoznysjc0rgq4c0su6l9z3

```console
$ rac -project test *.rac


2021/04/27 10:01:36 got stop packet without a start packet (Payload_MM_20210421-104323_20210422-133351.rac)
2021/04/27 10:01:37 Raw CCDImage Payload_MM_20210421-104323_20210422-133351_1301992928335388184_1.png has 1045504 pixels, but dimensions 2048 x 511
2021/04/27 10:01:37 Payload_MM_20210421-104323_20210422-133351_1301992928335388184_1.png: Found 1045504 pixels, but dimension 2048 x 511 says it should be 1046528
2021/04/27 10:01:38 got stop packet without a start packet (Payload_MM_20210421-104323_20210422-133351.rac)
2021/04/27 10:01:38 orphaned multi-package data without termination detected (Payload_MM_20210421-104323_20210422-133351.rac)
panic: runtime error: index out of range [2091008] with length 2091008

goroutine 1212 [running]:
image.(*Gray16).Gray16At(0xc003aae080, 0x400, 0x1fe, 0xc00044dcc6)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/image.go:783 +0xa2
image.(*Gray16).At(0xc003aae080, 0x400, 0x1fe, 0xbb2ee0, 0xc00044dcc6)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/image.go:775 +0x3f
image/png.(*encoder).writeImage(0xc000254f00, 0xbb1c20, 0xc00053a100, 0xbba540, 0xc003aae080, 0xc, 0xffffffffffffffff, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/png/writer.go:465 +0x1362
image/png.(*encoder).writeIDATs(0xc000254f00)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/png/writer.go:531 +0xe2
image/png.(*Encoder).Encode(0xc00000e0c0, 0xbb2540, 0xc000536028, 0xbba540, 0xc003aae080, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/png/writer.go:632 +0x371
image/png.Encode(...)
	/opt/hostedtoolcache/go/1.14.4/x64/src/image/png/writer.go:561
github.com/innosat-mats/rac-extract-payload/internal/exports.DiskCallbackFactory.func1.1(0xc000028c30, 0xc0004ba9f0, 0x7ffd6a04ceaf, 0x4, 0xc00022cee0)
	/home/runner/work/rac-extract-payload/rac-extract-payload/internal/exports/disk_writer.go:77 +0x1ce
created by github.com/innosat-mats/rac-extract-payload/internal/exports.DiskCallbackFactory.func1
	/home/runner/work/rac-extract-payload/rac-extract-payload/internal/exports/disk_writer.go:67 +0x1bb
```

--- a/go.mod
+++ b/go.mod
@@ -6,5 +6,5 @@ require (
 	github.com/aws/aws-sdk-go v1.31.0
 	github.com/howeyc/crc16 v0.0.0-20171223171357-2b2a61e366a6
 	github.com/jbuchbinder/gopnm v0.0.0-20150223212718-5176c556b9ce
-	golang.org/x/lint v0.0.0-20200302205851-738671d3881b
+	golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
 )

--- a/internal/exports/aws_s3_writer.go
+++ b/internal/exports/aws_s3_writer.go
@@ -67,7 +67,7 @@ func AWSS3CallbackFactory(
 		recoverWrite := func(imageFileName string) {
 			if r := recover(); r != nil {
 				log.Printf(
-					"Could not process image %s (%v)",
+					"Processing incomplete for image %s, skipping (%v)",
 					imageFileName, r,
 				)
 			}

--- a/internal/exports/aws_s3_writer.go
+++ b/internal/exports/aws_s3_writer.go
@@ -64,6 +64,14 @@ func AWSS3CallbackFactory(
 			log.Println(pkg.Error)
 			return
 		}
+		recoverWrite := func(imageFileName string) {
+			if r := recover(); r != nil {
+				log.Printf(
+					"Could not process image %s (%v)",
+					imageFileName, r,
+				)
+			}
+		}
 		if writeImages {
 			switch pkg.Data.(type) {
 			case *aez.CCDImage:
@@ -77,11 +85,14 @@ func AWSS3CallbackFactory(
 				go func() {
 					defer wg.Done()
 					imgFileName := ccdImage.FullImageName(project)
+					defer recoverWrite(imgFileName)
 					img := ccdImage.Image(pkg.Buffer)
 					pngBuffer := bytes.NewBuffer([]byte{})
-					png.Encode(pngBuffer, img)
+					err := png.Encode(pngBuffer, img)
+					if err != nil {
+						log.Panicf("failed encoding %s: %s", imgFileName, err)
+					}
 					upload(uploader, imgFileName, pngBuffer)
-
 					jsonFileName := GetJSONFilename(imgFileName)
 					jsonBuffer := bytes.NewBuffer([]byte{})
 					WriteJSON(jsonBuffer, &pkg, jsonFileName)

--- a/internal/exports/aws_s3_writer_test.go
+++ b/internal/exports/aws_s3_writer_test.go
@@ -103,6 +103,55 @@ func TestAWSS3CallbackFactory(t *testing.T) {
 			},
 		},
 		{
+			"Skips image with wrong image shape",
+			args{
+				project:     "myproj",
+				writeImages: true,
+			},
+			[]common.DataRecord{
+				{
+					Origin:         &common.OriginDescription{Name: "MyRac.rac"},
+					RamsesHeader:   &ramses.Ramses{},
+					RamsesTMHeader: &ramses.TMHeader{},
+					SourceHeader:   &innosat.SourcePacketHeader{},
+					TMHeader:       &innosat.TMHeader{},
+					RID:            aez.CCD5,
+					Data: &aez.CCDImage{
+						PackData: &aez.CCDImagePackData{
+							EXPTS: 5,
+							JPEGQ: aez.JPEGQUncompressed16bit,
+							NCOL:  42,
+							NROW:  42,
+						},
+						ImageFileName: "MyRac_wrong_shape_5.png",
+					},
+					Buffer: make([]byte, 2*2*2), // 2x2 pixels, 2 bytes per pix
+				},
+				{
+					Origin:         &common.OriginDescription{Name: "MyRac.rac"},
+					RamsesHeader:   &ramses.Ramses{},
+					RamsesTMHeader: &ramses.TMHeader{},
+					SourceHeader:   &innosat.SourcePacketHeader{},
+					TMHeader:       &innosat.TMHeader{},
+					RID:            aez.CCD5,
+					Data: &aez.CCDImage{
+						PackData: &aez.CCDImagePackData{
+							EXPTS: 5,
+							JPEGQ: aez.JPEGQUncompressed16bit,
+							NCOL:  1,
+							NROW:  2,
+						},
+						ImageFileName: "MyRac_5000000000_5.png",
+					},
+					Buffer: make([]byte, 2*2*2), // 2x2 pixels, 2 bytes per pix
+				},
+			},
+			map[string]int{
+				"myproj/MyRac_5000000000_5.png":  76,  // 8 + header
+				"myproj/MyRac_5000000000_5.json": 876, // length of the json
+			},
+		},
+		{
 			"Doesn't upload image when told not to",
 			args{
 				project:     "myproj",

--- a/internal/exports/disk_writer.go
+++ b/internal/exports/disk_writer.go
@@ -57,8 +57,8 @@ func DiskCallbackFactory(
 		recoverWrite := func(imageFileName string) {
 			if r := recover(); r != nil {
 				log.Printf(
-					"Could not process image %s (skipping...)",
-					imageFileName,
+					"Could not process image %s (%v)",
+					imageFileName, r,
 				)
 			}
 		}
@@ -75,24 +75,22 @@ func DiskCallbackFactory(
 				go func() {
 					defer wg.Done()
 					imgFileName := ccdImage.FullImageName(output)
-					// png.Encode is known to be able to panic, but recovery is
-					// defered here, in case something else also misbehave
 					defer recoverWrite(imgFileName)
 					img := ccdImage.Image(pkg.Buffer)
 					imgFile, err := os.Create(imgFileName)
 					if err != nil {
-						log.Fatalf("failed creating %s: %s", imgFileName, err)
+						log.Panicf("failed creating %s: %s", imgFileName, err)
 					}
 					defer imgFile.Close()
 					err = png.Encode(imgFile, img)
 					if err != nil {
-						log.Fatalf("failed encoding %s: %s", imgFileName, err)
+						log.Panicf("failed encoding %s: %s", imgFileName, err)
 					}
 					jsonFileName := GetJSONFilename(imgFileName)
 					jsonFile, err := os.Create(jsonFileName)
 					defer jsonFile.Close()
 					if err != nil {
-						log.Fatalf("failed creating %s: %s", jsonFileName, err)
+						log.Panicf("failed creating %s: %s", jsonFileName, err)
 					}
 					WriteJSON(jsonFile, &pkg, jsonFileName)
 				}()

--- a/internal/exports/disk_writer.go
+++ b/internal/exports/disk_writer.go
@@ -60,6 +60,8 @@ func DiskCallbackFactory(
 					"Could not process image %s (%v)",
 					imageFileName, r,
 				)
+				os.Remove(imageFileName)
+				os.Remove(GetJSONFilename(imageFileName))
 			}
 		}
 		if writeImages {

--- a/internal/exports/disk_writer.go
+++ b/internal/exports/disk_writer.go
@@ -57,7 +57,7 @@ func DiskCallbackFactory(
 		recoverWrite := func(imageFileName string) {
 			if r := recover(); r != nil {
 				log.Printf(
-					"Could not process image %s (%v)",
+					"Processing incomplete for image %s, skipping (%v)",
 					imageFileName, r,
 				)
 				os.Remove(imageFileName)

--- a/internal/exports/disk_writer.go
+++ b/internal/exports/disk_writer.go
@@ -54,6 +54,14 @@ func DiskCallbackFactory(
 		if pkg.Error != nil {
 			log.Println(pkg.Error)
 		}
+		recoverWrite := func(imageFileName string) {
+			if r := recover(); r != nil {
+				log.Printf(
+					"Could not process image %s (skipping...)",
+					imageFileName,
+				)
+			}
+		}
 		if writeImages {
 			switch pkg.Data.(type) {
 			case *aez.CCDImage:
@@ -67,21 +75,24 @@ func DiskCallbackFactory(
 				go func() {
 					defer wg.Done()
 					imgFileName := ccdImage.FullImageName(output)
+					// png.Encode is known to be able to panic, but recovery is
+					// defered here, in case something else also misbehave
+					defer recoverWrite(imgFileName)
 					img := ccdImage.Image(pkg.Buffer)
 					imgFile, err := os.Create(imgFileName)
 					if err != nil {
-						log.Printf("failed creating %s: %s", imgFileName, err)
-						panic(err.Error())
+						log.Fatalf("failed creating %s: %s", imgFileName, err)
 					}
 					defer imgFile.Close()
-					png.Encode(imgFile, img)
-
+					err = png.Encode(imgFile, img)
+					if err != nil {
+						log.Fatalf("failed encoding %s: %s", imgFileName, err)
+					}
 					jsonFileName := GetJSONFilename(imgFileName)
 					jsonFile, err := os.Create(jsonFileName)
 					defer jsonFile.Close()
 					if err != nil {
-						log.Printf("failed creating %s: %s", jsonFileName, err)
-						panic(err.Error())
+						log.Fatalf("failed creating %s: %s", jsonFileName, err)
 					}
 					WriteJSON(jsonFile, &pkg, jsonFileName)
 				}()

--- a/internal/exports/disk_writer_test.go
+++ b/internal/exports/disk_writer_test.go
@@ -273,7 +273,7 @@ func TestDiskCallbackFactoryCreator(t *testing.T) {
 			},
 		},
 		{
-			"Continues on error",
+			"Continues on error due to wrong image shape",
 			args{writeImages: true},
 			[]common.DataRecord{
 				{
@@ -286,7 +286,7 @@ func TestDiskCallbackFactoryCreator(t *testing.T) {
 							NROW:  42,
 							EXPTS: 5,
 						},
-						ImageFileName: "File1_5000000000_2.png",
+						ImageFileName: "File1_wrong_shape_2.png",
 					},
 					Buffer: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
 				},
@@ -306,7 +306,7 @@ func TestDiskCallbackFactoryCreator(t *testing.T) {
 				},
 			},
 			[]wantFile{
-				{"File1_5000000000_2.png", 0, true},
+				{"File1_wrong_shape_2.png", 0, true},
 				{"File1_6000000000_3.png", 0, true},
 				{"File1_6000000000_3.json", 0, true},
 			},

--- a/internal/exports/disk_writer_test.go
+++ b/internal/exports/disk_writer_test.go
@@ -273,6 +273,45 @@ func TestDiskCallbackFactoryCreator(t *testing.T) {
 			},
 		},
 		{
+			"Continues on error",
+			args{writeImages: true},
+			[]common.DataRecord{
+				{
+					RID:    aez.CCD2,
+					Origin: &common.OriginDescription{Name: "File1.rac"},
+					Data: &aez.CCDImage{
+						PackData: &aez.CCDImagePackData{
+							JPEGQ: aez.JPEGQUncompressed16bit,
+							NCOL:  42,
+							NROW:  42,
+							EXPTS: 5,
+						},
+						ImageFileName: "File1_5000000000_2.png",
+					},
+					Buffer: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+				},
+				{
+					RID:    aez.CCD3,
+					Origin: &common.OriginDescription{Name: "File1.rac"},
+					Data: &aez.CCDImage{
+						PackData: &aez.CCDImagePackData{
+							JPEGQ: aez.JPEGQUncompressed16bit,
+							NCOL:  1,
+							NROW:  2,
+							EXPTS: 6,
+						},
+						ImageFileName: "File1_6000000000_3.png",
+					},
+					Buffer: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+				},
+			},
+			[]wantFile{
+				{"File1_5000000000_2.png", 0, true},
+				{"File1_6000000000_3.png", 0, true},
+				{"File1_6000000000_3.json", 0, true},
+			},
+		},
+		{
 			"Doesn't creates images when asked not to",
 			args{writeImages: false},
 			[]common.DataRecord{

--- a/internal/exports/disk_writer_test.go
+++ b/internal/exports/disk_writer_test.go
@@ -306,7 +306,6 @@ func TestDiskCallbackFactoryCreator(t *testing.T) {
 				},
 			},
 			[]wantFile{
-				{"File1_wrong_shape_2.png", 0, true},
 				{"File1_6000000000_3.png", 0, true},
 				{"File1_6000000000_3.json", 0, true},
 			},

